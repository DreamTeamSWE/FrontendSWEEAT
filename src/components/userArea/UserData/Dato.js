import React from 'react';

function Dato(props) {
    return <div className="dato">
        <label className="dateName" >{props.dateName}</label>
        <label className="dateValue">{props.dateValue}</label>
    </div>
}
export default Dato;
