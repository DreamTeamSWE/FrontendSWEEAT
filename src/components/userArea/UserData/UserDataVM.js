import { useState, useEffect } from "react";

import { useForm } from "react-hook-form";

export default function DatiPersonaliViewModel(store) {
  //Model
  const userStore = store.userStore;

  //No need for observer. It is supposed that the user is logged when you access this area.
  useEffect(() => {
    setUserData(userStore.userDetails);
  }, []);

  const {
    register,
    formState: { errors },
    handleSubmit,
    getValues,
  } = useForm({ criteriaMode: "all" });

  const [modifica, setModifica] = useState(false);
  const [stato, setStato] = useState("inizio");
  const [userData, setUserData] = useState("");
  const changeState = () => {
    setModifica(!modifica);
  };
  async function cambiaPasswordHandler(data) {
    const { oldPW, newPW } = data;
    const success = await userStore.changePW(oldPW, newPW);
    if (success) setStato("cambiato");
    else setStato("errore");
  }

  return {
    register,
    errors,
    handleSubmit,
    getValues,
    modifica,
    stato,
    userData,
    changeState,
    cambiaPasswordHandler,
  };
}
