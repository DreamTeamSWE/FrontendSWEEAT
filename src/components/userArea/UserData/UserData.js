import Dato from "./Dato";
import React from "react";

import { ErrorMessage } from "@hookform/error-message";
import DatiPersonaliViewModel from "./UserDataVM";
import { useStore } from "../../../Context";
import "./../../auth/css/formStyle.css";

function DatiPersonali() {
  const {
    register,
    errors,
    handleSubmit,
    getValues,
    modifica,
    stato,
    userData,
    changeState,
    cambiaPasswordHandler,
  } = DatiPersonaliViewModel(useStore());

  return (
    <>
      {modifica && (
        <div className="datiPersonali">
          <h1>Modifica password</h1>
          <form className="form" onSubmit={handleSubmit(cambiaPasswordHandler)}>
            <div className="field">
              <label>Password attuale</label>
              <input
                {...register("oldPW", {
                  required: "Password attuale necessaria",
                  pattern: {
                    value:
                      /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/u,
                    message:
                      "La password deve essere lunga almeno 8 caratteri e contenere almeno una lettera maiuscola, una minuscola, un numero e un simbolo tra @$!%*?&.",
                  },
                })}
                type="password"
                placeholder="Password attuale"
              />
              <ErrorMessage
                errors={errors}
                name="oldPW"
                render={({ message }) => (
                  <p className="messaggioErrore">{message}</p>
                )}
              />
            </div>
            <div className="field">
              <label>Password nuova</label>
              <input
                {...register("newPW", {
                  required: "Password nuova necessaria",
                  pattern: {
                    value:
                      /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/u,
                    message:
                      "La password deve essere lunga almeno 8 caratteri e contenere almeno una lettera maiuscola, una minuscola, un numero e un simbolo tra @$!%*?&.",
                  },
                  validate: (value) =>
                    value !== getValues("oldPW") ||
                    "La nuova password deve essere diversa da quella precedente.",
                })}
                type="password"
                placeholder="Nuova password"
              />
              <ErrorMessage
                errors={errors}
                name="newPW"
                render={({ message }) => (
                  <p className="messaggioErrore">{message}</p>
                )}
              />
            </div>
            <div className="button">
              <input type="submit" value="Cambia password" />
            </div>
            {stato !== "inizio" ? (
              stato === "cambiato" ? (
                <label>Password cambiata correttamente</label>
              ) : (
                <label>Password non cambiata</label>
              )
            ) : null}
          </form>
        </div>
      )}
      {!modifica && (
        <div className="datiPersonali">
          <h1>Dati personali</h1>
          {userData ? (
            <div>
              <Dato
                key={userData.surname}
                dateName="Cognome"
                dateValue={userData.surname}
              />
              <Dato
                key={userData.name}
                dateName="Nome"
                dateValue={userData.name}
              />
              <Dato
                key={userData.email}
                dateName="Email"
                dateValue={userData.email}
              />
              <button onClick={changeState}>Modifica password</button>
            </div>
          ) : null}
        </div>
      )}
    </>
  );
}

export default DatiPersonali;
