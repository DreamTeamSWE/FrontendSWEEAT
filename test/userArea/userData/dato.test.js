import Dato from '../../../src/components/userArea/UserData/Dato';
import UserData from '../../../src/components/userArea/UserData/UserData';
import '@testing-library/jest-dom';
import React from 'react';
import {configure, shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {render, screen} from '@testing-library/react';
import {BrowserRouter as Router} from 'react-router-dom';

configure({adapter: new Adapter()});

test('Check Dato', () =>{
  const dato = shallow(<Dato/>);
  expect(dato).toMatchSnapshot();;
});

test('Check User Data', () =>{
  const ud = shallow(<UserData/>);
  expect(ud).toMatchSnapshot();;
});
