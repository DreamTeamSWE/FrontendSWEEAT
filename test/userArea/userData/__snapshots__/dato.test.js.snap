// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Dato 1`] = `ShallowWrapper {}`;

exports[`Check User Data 1`] = `ShallowWrapper {}`;
